src/lib.rs

```diff
@@ -74,87 +73,60 @@ impl<I> nom::error::ParseError<I> for Error {
   }
 }
 
+pub trait DlmsDataLinkLayer<I, O, F> {
```

Comment:
I think this trait can be simplified a bit. In the current implementations `I` and `O` is always the same. `F` is either `Cow<'_, [u8]>` or `Vec<u8>`, I think this could be fixed to `Cow<'_, [u8]>` in all cases.

Also, currently, we have two new struct types which implement this trait, I think it would make sense to directly implement the trait for the telegram types themselves, i.e. `impl<'i, 'f> DataLinkLayer for &'i [mbusparse::Telegram<'f>]`.

Reply:
The two structs/enums are needed to get rid of the lifetimes from `Telegram<'i>` and `HdlcFrame<'i>`. This is important for `smart_meter` since these lifetimes do not exist any more. See [smart-meter-rs#10](https://github.com/reitermarkus/smart-meter-rs/pull/10) for more details.